service/history/workflow/workflow_task_state_machine.go

```diff
@@ -689,41 +690,55 @@ func (m *workflowTaskStateMachine) GetWorkflowTaskInfo(
-		Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
-	}
-
-	scheduledEvent := &historypb.HistoryEvent{
-		EventId:   workflowTask.ScheduledEventID,
-		EventTime: workflowTask.ScheduledTime,
-		EventType: enumspb.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED,
-		Version:   m.ms.currentVersion,
-		Attributes: &historypb.HistoryEvent_WorkflowTaskScheduledEventAttributes{
-			WorkflowTaskScheduledEventAttributes: &historypb.WorkflowTaskScheduledEventAttributes{
-				TaskQueue:           taskQueue,
-				StartToCloseTimeout: workflowTask.WorkflowTaskTimeout,
-				Attempt:             workflowTask.Attempt,
+) *historyspb.TransientWorkflowTaskInfo {
+
+	var historySuffix []*historypb.HistoryEvent
+
+	var transientWorkflowTask *historyspb.TransientWorkflowTaskInfo
+	if workflowTask.Attempt > 1 {
+		// This workflowTask is retried from mutable state
+		// Also return schedule and started which are not written to history yet
+		scheduledEvent := &historypb.HistoryEvent{
+			EventId:   workflowTask.ScheduledEventID,
+			EventTime: workflowTask.ScheduledTime,
+			EventType: enumspb.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED,
+			Version:   m.ms.currentVersion,
+			Attributes: &historypb.HistoryEvent_WorkflowTaskScheduledEventAttributes{
+				WorkflowTaskScheduledEventAttributes: &historypb.WorkflowTaskScheduledEventAttributes{
+					TaskQueue: &taskqueuepb.TaskQueue{
+						Name: m.ms.executionInfo.TaskQueue,
+						Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
+					},
+					StartToCloseTimeout: workflowTask.WorkflowTaskTimeout,
+					Attempt:             workflowTask.Attempt,
+				},
 			},
-		},
-	}
-
-	startEvent := &historypb.HistoryEvent{
-		EventId:   workflowTask.StartedEventID,
-		EventTime: workflowTask.StartedTime,
-		EventType: enumspb.EVENT_TYPE_WORKFLOW_TASK_STARTED,
-		Version:   m.ms.currentVersion,
-		Attributes: &historypb.HistoryEvent_WorkflowTaskStartedEventAttributes{
-			WorkflowTaskStartedEventAttributes: &historypb.WorkflowTaskStartedEventAttributes{
-				ScheduledEventId: workflowTask.ScheduledEventID,
-				Identity:         identity,
-				RequestId:        workflowTask.RequestID,
+		}
+
+		startedEvent := &historypb.HistoryEvent{
+			EventId:   workflowTask.StartedEventID,
+			EventTime: workflowTask.StartedTime,
+			EventType: enumspb.EVENT_TYPE_WORKFLOW_TASK_STARTED,
+			Version:   m.ms.currentVersion,
+			Attributes: &historypb.HistoryEvent_WorkflowTaskStartedEventAttributes{
+				WorkflowTaskStartedEventAttributes: &historypb.WorkflowTaskStartedEventAttributes{
+					ScheduledEventId: workflowTask.ScheduledEventID,
+					Identity:         identity,
+					RequestId:        workflowTask.RequestID,
+				},
 			},
-		},
+		}
+
+		// TODO (mmcshane): remove population of ScheduledEvent and StartedEvent
+		// after v1.18 is released
+		transientWorkflowTask = &historyspb.TransientWorkflowTaskInfo{
+			ScheduledEvent: scheduledEvent,
+			StartedEvent:   startedEvent,
+			HistorySuffix:  append(historySuffix, scheduledEvent, startedEvent),
```

Comment:
Slightly weird to have the `historySuffix` variable in the outer scope - I would have expected just `HistorySuffix: []*history.Event{scheduledEvent, startedEvent}` here but I assume the upcoming update implementation will make use of the `historySuffix`.

Reply:
Yes, exactly. It was cherry-picked from another commit where it does make sense.